### PR TITLE
stats: In StatNameSet, differentiate between dynamic and builtin name lookup, which should have a fallback and avoid  locks

### PR DESCRIPTION
**Pull request by jmarantz**
Description: In #7890 it became clear we could enumerate the set of possible Redis commands, and use a fallback symbol to report stats on unexpected redis commands, to avoid taking an unexpected lock. This captures that semantic in StatNameSet. Unfortunately there are still some classes of stat name lookups that appear to be dynamic, so will generally require locks. This PR calls them out explicitly and we can potentially discuss them individually in this PR and decide what we want to do.
Risk Level: low
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
Looks good to me; appreciate the simplification.

File: source/extensions/filters/http/dynamo/dynamo_stats.cc [hunk 1]
```diff
@@ -37,6 +39,11 @@ DynamoStats::DynamoStats(Stats::Scope& scope, const std::string& prefix)
   }
   RequestParser::forEachStatString(
       [this](const std::string& str) { stat_name_set_.rememberBuiltin(str); });
+  for (int32_t status_code : {200, 400, 403, 502}) {
```
Comment (mattklein123) [inline source/extensions/filters/http/dynamo/dynamo_stats.cc]: nit: unsigned

File: source/common/stats/symbol_table_impl.cc [hunk 2]
```diff
@@ -442,19 +442,35 @@ void StatNameSet::rememberBuiltin(absl::string_view str) {
   builtin_stat_names_[str] = stat_name;
 }
 
-Stats::StatName StatNameSet::getStatName(absl::string_view token) {
+StatName StatNameSet::getBuiltin(absl::string_view token, StatName fallback) {
+  StatName stat_name = getBuiltinHelper(token, fallback);
+  if (stat_name == fallback) {
```
Comment (mattklein123) [inline source/common/stats/symbol_table_impl.cc]: perf nit: you could likely avoid the comparison here by potentially returning a struct from the helper (or something like that) as to whether it selected the fallback?

File: source/common/stats/symbol_table_impl.cc [hunk 3]
```diff
@@ -442,19 +442,35 @@ void StatNameSet::rememberBuiltin(absl::string_view str) {
...
+    // TODO(#8116): do periodic logging here.
+    // ENVOY_LOG_MISC(warn, "Builtin stat name not pre-registered: {}", token);
```
Comment (mattklein123) [inline source/common/stats/symbol_table_impl.cc]: I'm still skeptical that the logging is the best path forward. What about doing something like:
1) Increment some stat when we hit this path.
2) Possible register the name with some admin handler, which could then be dumped by an operator to debug?

**Comment by jmarantz** [comment]
I'll address the code nits later today but wanted to carry on the dialog.

I would love to not have any dynamic stats. I handled HTTP status (I think) acceptably via an array of atomics in https://github.com/envoyproxy/envoy/blob/81460d8482b73ba70ea15b76163a7c645e2a7b96/source/common/http/codes.cc#L186 and I could factor that out and use for other cases where there's a status-like thing that's numerically bounded.

I could re-do dynamo_filter.cc that way if the 'status' reported there is really an http-style status as opposed to any arbitrary integer. I don't know anything about Dynamo so guidance here would be appreciated.

We also name stats for tables, which I'm guessing are totally data-driven in dynamo_stats.cc, so removing support for getDynamic() would be a regression there. I'm not sure if there is also a known set of choices for error_type.  Same with partition-id.

source/extensions/filters/http/fault/fault_filter.cc names a stat for the downstream cluster. We could plumb that all the way up and symbolize a stat-name for that when the downstream cluster is created. That probably wouldn't be too bad.

source/extensions/filters/http/ip_tagging/ip_tagging_filter.cc increments a stat based on a tag. I'm not sure if/when we can enumerate the legal set of tags.

mongo proxy names stats for collections and call-sites. Can/should we enumerate those?

Then there's the wasm folks (@mandarjog et al) who are creating named stats from wasm code. But I think that might be OK because the flow from wasm is that you allocate a stat by name (taking a lock no matter what if that stat was never seen before) and then get a handle to it which you can re-use in your filter. So that might be OK?

**Comment by mattklein123** [comment]
Hmm yeah I don't think we are going to easily be able to get rid of dynamics right now, but I agree our default policy should be to not use them where it makes sense to do so. Thanks for the very detailed explanation.

File: source/common/stats/symbol_table_impl.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/stats/symbol_table_impl.cc]: Sorry I realize you are doing something like this in the other PR. I will review that today and then we can circle back to this call-site.

Comment (jmarantz) [reply source/common/stats/symbol_table_impl.cc]: I removed the commented-out code and the helper function, per the discussion on Slack where you convinced me not to add logging initially.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
For a quick sanity check I built the PR against my test environment and verified the stats for redis look good.

**Comment by jmarantz** [comment]
Thanks Nicolas!

File: source/common/stats/symbol_table_impl.h [hunk 4]
```diff
@@ -671,15 +683,19 @@ class StatNameSet {
    * set's mutex and also the SymbolTable mutex which must be taken during
    * StatNamePool::add().
    */
-  StatName getStatName(absl::string_view token);
+  StatName getDynamic(absl::string_view token);
+  StatName getBuiltin(absl::string_view token, StatName fallback);
```
Comment (antoniovicente) [inline source/common/stats/symbol_table_impl.h]: Docstring?

**Pull request merged**